### PR TITLE
Switch to PSR-17 and remove deprecated code

### PR DESCRIPTION
**Pull request by GrahamCampbell**
This PR removes all the deprecated code from the codebase, and switches to PSR-17 which replaces the old PHP HTTP interfaces, just like PSR-18 replaced some of the others (implemented in an earlier PR).

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Bahaha... I'm guessing this is not working:

image

File: doc/security.md [hunk 1]
```diff
@@ -15,48 +16,29 @@ $client->authenticate($usernameOrToken, $password, $method);
 `$usernameOrToken` is, of course, the username (or in some cases token/client ID, more details you can find below),
 and guess what should contain `$password`. The `$method` can contain one of the five allowed values:
```
Comment (cs278) [inline doc/security.md:17]:
```suggestion
and guess what should contain `$password`. The `$method` can contain one of the three allowed values:
```

File: lib/Github/HttpClient/Plugin/Authentication.php [hunk 2]
```diff
@@ -46,62 +46,26 @@ public function __construct($tokenOrLogin, $password, $method)
...
+    {
+        $request = $request->withHeader(
+            'Authorization',
+            $this->getAuthorizationHeader()
+        );
+
+        return $next($request);
+    }
+
+    private function getAuthorizationHeader()
```
Comment (acrobat) [inline lib/Github/HttpClient/Plugin/Authentication.php:58]: Add a `string` return type for this method

File: test/Github/Tests/Api/RateLimitTest.php [hunk 3]
```diff
@@ -37,11 +37,9 @@ class RateLimitTest extends TestCase
     protected $api;
 
     /**
-     * Used to construct common expectations for the API input data in each unit test.
-     *
-     * {@inheritdoc}
+     * @before
      */
-    protected function setUp(): void
+    public function initMocks()
```
Comment (acrobat) [inline test/Github/Tests/Api/RateLimitTest.php:42]: Just curious why this change and not just keep the setUp?

File: lib/Github/Client.php [hunk 4]
```diff
@@ -336,17 +305,13 @@ public function api($name)
      */
     public function authenticate($tokenOrLogin, $password = null, $authMethod = null)
     {
-        if (null === $password && null === $authMethod) {
-            throw new InvalidArgumentException('You need to specify authentication method!');
-        }
-
-        if (null === $authMethod && in_array($password, [self::AUTH_URL_TOKEN, self::AUTH_URL_CLIENT_ID, self::AUTH_HTTP_PASSWORD, self::AUTH_HTTP_TOKEN, self::AUTH_JWT], true)) {
```
Comment (acrobat) [inline lib/Github/Client.php:343]: Shouldn't we keep the `in_array` check to verifiy users provide a supported method?

**Comment by acrobat** [comment]
Thanks for the work already @GrahamCampbell! Just a few comments/changes.

File: test/Github/Tests/Api/RateLimitTest.php (hunk 3 above)
Comment (GrahamCampbell) [reply test/Github/Tests/Api/RateLimitTest.php:42]: So we don't get messed about by phpunit's own type signatures. It's always a pain when upgrading.

File: lib/Github/HttpClient/Plugin/Authentication.php (hunk 2 above)
Comment (GrahamCampbell) [reply lib/Github/HttpClient/Plugin/Authentication.php:58]: Do you want a genuine return type, or just phpdoc. Let me know, and I will add the one you want. :)

File: lib/Github/Client.php (hunk 4 above)
Comment (GrahamCampbell) [reply lib/Github/Client.php:343]: Yeh, something's got lost in the refactoring here. I'll take a look. :)

File: lib/Github/HttpClient/Plugin/Authentication.php (hunk 2 above)
Comment (acrobat) [reply lib/Github/HttpClient/Plugin/Authentication.php:58]: I would add a "real" return type here

File: test/Github/Tests/Api/RateLimitTest.php (hunk 3 above)
Comment (acrobat) [reply test/Github/Tests/Api/RateLimitTest.php:42]: Good 👍

**Comment by GrahamCampbell** [comment]
Just rebased and applied changes based on the feedback.

**Comment by GrahamCampbell** [comment]
@acrobat If you do another 2.x -> master merge, I'll rebase this PR again, so that you don't have additional merge conflicts in the authentication class later on.

**Comment by acrobat** [comment]
Upmerge is done!

**Comment by GrahamCampbell** [comment]
Build green now. :)

**Comment by acrobat** [comment]
Thanks a lot for your work on this @GrahamCampbell!

**Comment by GrahamCampbell** [comment]
No problem. :)
